### PR TITLE
Clean up bufferToArray()

### PR DESCRIPTION
**Pull request by tad88dev**
- `readUInt32BE(i * 4)` throw out of range error if `Buffer.length` isn't multiples of 4
- `buf.length / 4` Is a `float` number, it has been converted to `int` by `(buf.length / 4) | 0`
- Use for loop instead of while loop

### CODE REVIEW

**Comment by fanatid** [comment]
`unit_max` optimization is useless because `Math.pow` called only once on initialization

File: aes.js [hunk 1]
```diff
@@ -81,14 +81,17 @@ AES.keySize = 256 / 8
...
 function bufferToArray (buf) {
-  var len = buf.length / 4
+  if (!Buffer.isBuffer(buf)) {
+    throw new TypeError('Input must be a Buffer')
+  }
+  var len = (buf.length / 4) | 0
   var out = new Array(len)
-  var i = -1
-  while (++i < len) {
+  for (var i = 0; i < len; i++) {
```
Comment (dcousens) [inline aes.js]: ACK this, and only this, IMHO

**Comment by dcousens** [comment]
ACK c740b5f8cfbdae9e8b2e3bdf08ea757a67bcd762, NACK fca8110

File: aes.js (hunk 1 above)
Comment (calvinmetcalf) [inline aes.js]: the while loop idiom is one I use all over my code due to my distrust of for loops so probably not worth changing on it's own

File: aes.js [hunk 2]
```diff
@@ -81,14 +81,17 @@ AES.keySize = 256 / 8
 AES.prototype.keySize = AES.keySize
 
 function bufferToArray (buf) {
-  var len = buf.length / 4
+  if (!Buffer.isBuffer(buf)) {
```
Comment (calvinmetcalf) [inline aes.js]: I'm pretty sure bufferToArray is an internal function so it may make more sense to put the check outside the function because we may be calling it in situations where we are confident that it is a buffer

File: aes.js [hunk 3]
```diff
@@ -81,14 +81,17 @@ AES.keySize = 256 / 8
...
+    throw new TypeError('Input must be a Buffer')
+  }
+  var len = (buf.length / 4) | 0
```
Comment (calvinmetcalf) [inline aes.js]: can we have a comment here about the exact behavior of this? it converts to int32 via truncation right ?

**Comment by calvinmetcalf** [comment]
> Math.pow(2,32) x 64,460,793 ops/sec ±0.78% (89 runs sampled)
> 0xffffffff + 1 x 65,468,118 ops/sec ±0.65% (90 runs sampled)
> Fastest is 0xffffffff + 1

these are so close to the margin of error as to be functionally identical, don't bother

File: aes.js (hunk 3 above)
Comment (tad88dev) [inline aes.js]: Ok, i will put a comment at this line.

File: aes.js (hunk 1 above)
Comment (tad88dev) [inline aes.js]: It's make source code more clear and readable.

Comment (calvinmetcalf) [inline aes.js]: at the very least this should be pulled out into a separate commit or pull as it's totally unrelated to the other thing

**Comment by tad88dev** [comment]
What's ddda5c1 ? : 
- fca8110 has beend truncated.
- Remove Buffer check
- Add comment

**Comment by dcousens** [comment]
> readUInt32BE(i * 4) throw out of range error if Buffer.length isn't multiples of 4

Was this even possible to hit?

**Comment by tad88dev** [comment]
@dcousens I have some inspect, look like `bufferToArray()` isn't callable from external. It's also pass the integrated test, but i'm not sure about unit test. Do we need to build the unit test?.

**Comment by dcousens** [comment]
@tad88dev I'm just asking,  because,  how did you come to the title:  "Fix out of range error" if in fact, that error is **never** hit?

**Comment by tad88dev** [comment]
Becuase i'm thinking like a machine :dancing_men:. Thanks, that was my bad.

**Comment by calvinmetcalf** [comment]
so this code is only ever used by keys passed to the AES class that is called only after [this code](https://github.com/crypto-browserify/browserify-aes/blob/7aca10974e0bbf34bcc009b29a6a9b4252d99c9d/encrypter.js#L99-L101) (key is refereed to as password here) meaning there is no way for a non divisible by 4 buffer to get passed here

**Comment by tad88dev** [comment]
@calvinmetcalf So, do you mean this PR unnecessary?

**Comment by calvinmetcalf** [comment]
yes that particular aspect is, maybe we should close this and you can add a new one that just aims for cleaning up my ... idiosyncratic ... code style

**Comment by tad88dev** [comment]
@calvinmetcalf Thanks, got it!.
